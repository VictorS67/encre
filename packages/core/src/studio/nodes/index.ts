import { type RecordId } from '../../load/keymap.js';
import { type Serializable } from '../../load/serializable.js';
import { type Callable } from '../../record/index.js';
import { type DataType, type ValueOf } from '../data.js';
import { type UIContext } from '../ui.js';

/**
 * Maps port names to their respective data types within a node.
 */
export type NodePortFields = {
  [key: string]: DataType | Readonly<DataType[]>;
};

/**
 * Maps port names to their respective data sizes, used to handle the flow or
 * limitations of data through each port.
 */
export type NodePortSizes = {
  [key: string]: number;
};

/**
 * Represents the basic structure of a node within a graphical or data flow environment.
 * This node carries both visual and functional properties.
 */
export interface BaseNode {
  /**
   * Unique identifier for the node.
   */
  id: RecordId;

  /**
   * Visual presentation details of the node, including position and size.
   */
  visualInfo: {
    position: {
      x: number;
      y: number;
      zIndex?: number;
    };
    size: {
      width: number;
      height: number;
    };
    content?: {
      color?:
        | 'red'
        | 'orange'
        | 'gold'
        | 'yellow'
        | 'palmera'
        | 'green'
        | 'meadow'
        | 'cyan'
        | 'blue'
        | 'cornflower'
        | 'purple'
        | 'pink'
        | 'razzmatazz'
        | 'silver'
        | 'dark';
    };
  };

  /**
   * Input ports for the node, mapping port names to data types.
   */
  inputs: NodePortFields | undefined;

  /**
   * Output ports for the node, mapping port names to data types.
   */
  outputs: NodePortFields | undefined;

  /**
   * Optional title for the node.
   */
  title?: string | undefined;

  /**
   * Optional description of the node's functionality.
   */
  description?: string | undefined;

  /**
   * Optional mapping of output ports to their respective data sizes.
   * Default is 0.
   */
  outputSizes?: NodePortSizes | undefined;
}

/**
 * Extends `BaseNode` to include serialization capabilities along with additional
 * runtime properties.
 */
export interface SerializableNode<
  NodeType extends string = string,
  NodeData extends Serializable = Serializable,
> extends BaseNode {
  /**
   * Type of the node, typically defining its category or functionality.
   */
  type: NodeType;

  /**
   * Subtype of the node, providing further specification within its type.
   */
  subType: string;

  /**
   * Optional arguments used to register or initialize the node.
   */
  registerArgs?: Record<string, unknown>;

  /**
   * Estimated runtime per process in milliseconds.
   */
  runtime?: number;

  /**
   * Estimated RAM usage per process in megabytes.
   */
  memory?: number;

  /**
   * Data associated with the node, which must be serializable.
   */
  data: NodeData;
}

/**
 * Specialized version of `SerializableNode` that represents nodes capable of performing
 * callable actions.
 */
export interface CallableNode<
  NodeType extends string = string,
  NodeData extends Callable = Callable,
> extends SerializableNode<NodeType, NodeData> {}

/**
 * Defines a node port, which can be either an input or output port, including details
 * about data types, requirements, and defaults.
 */
export type NodePortDef = NodeInputPortDef | NodeOutputPortDef;

/**
 * Defines the structure of an input port on a node. It specifies the port's data type,
 * requirements, and optional default values.
 */
export type NodeInputPortDef = {
  /**
   * Identifier for the node to which this port belongs.
   */
  nodeId: RecordId;

  /**
   * The name of the port.
   */
  name: string;

  /**
   * The data type or types that this port accepts.
   */
  type: DataType | Readonly<DataType[]>;

  /**
   * Indicates whether this port must be connected.
   */
  required?: boolean;

  /**
   * Stored data that can be provided to the port.
   */
  data?: ValueOf<DataType> | Readonly<ValueOf<DataType>[]>;

  /**
   * A default value for the port, used if no input is provided.
   */
  default?: unknown;
};

/**
 * Defines the structure of an output port on a node. It details the port's data type,
 * optional requirements, and possible default values.
 */
export type NodeOutputPortDef = {
  /**
   * Identifier for the node to which this port belongs.
   */
  nodeId: RecordId;

  /**
   * The name of the port.
   */
  name: string;

  /**
   * The data type or types that this port outputs.
   */
  type: DataType | Readonly<DataType[]>;

  /**
   * Indicates whether this port must be connected.
   */
  required?: boolean;

  /**
   * Stored data associated with the port, potentially for initial output or defaults.
   */
  data?: ValueOf<DataType> | Readonly<ValueOf<DataType>[]>;

  /**
   * A default value for the port, used if the output is unspecified.
   */
  default?: unknown;
};

/**
 * Represents a connection between two node ports, specifying the source and destination nodes
 * and ports.
 * 
 * For example, we have connection: A.portA ---> B.portB
 *
 * Then, we should have:
 * - fromNodeId: A
 * - fromPortName: portA
 * - toNodeId: B
 * - toPortName: portB
 */
export type NodeConnection = {
  /**
   * The identifier of the node from which the connection originates.
   */
  fromNodeId: RecordId;

  /**
   * The name of the port from which the connection originates.
   */
  fromPortName: string;

  /**
   * The identifier of the node to which the connection is made.
   */
  toNodeId: RecordId;

  /**
   * The name of the port to which the connection is made.
   */
  toPortName: string;
};

/**
 * Represents various forms of body content that can be associated with a node,
 * such as strings or UI context objects.
 */
export type NodeBody = string | UIContext | UIContext[] | undefined;
