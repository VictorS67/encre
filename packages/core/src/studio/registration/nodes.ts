import { load } from '../../load/index.js';
import { globalImportMap, globalSecretMap } from '../../load/registration.js';
import { Serializable } from '../../load/serializable.js';
import { type NodeImpl } from '../nodes/base.js';
import { type SerializableNode } from '../nodes/index.js';
import {
  GeminiChatNodeImpl,
  OpenAIChatNodeImpl,
} from '../nodes/inference/chat/chatlm.node.js';
import {
  GeminiNodeImpl,
  OpenAINodeImpl,
} from '../nodes/inference/chat/llm.node.js';
import { VariableValidatorNodeImpl } from '../nodes/inference/validate/validator.node.js';
import { PDFLoaderNodeImpl } from '../nodes/input/loader.node.js';
import {
  BotMessageNodeImpl,
  ChatMessageNodeImpl,
  FunctionMessageNodeImpl,
  HumanMessageNodeImpl,
  SystemMessageNodeImpl,
} from '../nodes/input/message.node.js';
import {
  ChatPromptNodeImpl,
  StringPromptNodeImpl,
} from '../nodes/input/prompt.node.js';
import {
  LanguageTextSplitterNodeImpl,
  RecursiveTextSplitterNodeImpl,
  TextSplitterNodeImpl,
  TokenTextSplitterNodeImpl,
} from '../nodes/input/splitter.node.js';
import { IfConditionNodeImpl } from '../nodes/utility/if.node.js';
import { InputNodeImpl } from '../nodes/utility/input.node.js';

type ExtractType<T> = T extends `${infer U}-${any}` ? U : never;

type ExtractSubType<T> = T extends `${any}-${infer U}` ? U : never;

/**
 * Provides a constructor interface for node implementations,
 * allowing creation of node instances and initialization from raw data.
 */
export interface NodeImplConstructor<T extends SerializableNode> {
  /**
   * Constructor that initializes with a node instance.
   */
  new (node: T): NodeImpl<T>;

  /**
   * Factory method to create a node with optional arguments.
   * @param args registry arguments for creating the specific node.
   */
  create(args?: Record<string, unknown>): T;

  /**
   * Initializes a node from a serializable object with optional arguments.
   * @param raw customized serializable to attach to the node's data.
   * @param args registry arguments for creating the specific node.
   */
  nodeFrom(raw: Serializable, args?: Record<string, unknown>): T;
}

/**
 * Manages registration and creation of various node types within a graph system,
 * supporting dynamic type and subtype management.
 */
export class NodeRegistration<
  NodeTypes extends string = never,
  NodeSubTypes extends string = never,
  Nodes extends SerializableNode = never,
> {
  /**
   * Stores instances of registered nodes.
   *
   * @internal
   */
  registeredNodes: Nodes = undefined!;

  /**
   * Tracks registered primary node types.
   *
   * @internal
   */
  registeredNodeTypes: NodeTypes = undefined!;

  /**
   * Maintains a mapping from node types to their subtypes.
   *
   * @internal
   */
  registeredNodeTypePairs: { [P in NodeTypes]: NodeSubTypes[] } = undefined!;

  /**
   * Stores implementation details for node types and subtypes.
   */
  info = {} as {
    [P in `${NodeTypes}-${NodeSubTypes}`]: {
      impl: NodeImplConstructor<
        Extract<Nodes, { type: ExtractType<P>; subType: ExtractSubType<P> }>
      >;
    };
  };

  /**
   * Maps type-subtype pairs to their corresponding node implementations and initialization data.
   */
  implsMap = {} as Record<
    string,
    { impl: NodeImplConstructor<SerializableNode>; init: SerializableNode }
  >;

  /**
   * Maps serializable IDs to their node implementations for nodes that are initialized from raw data.
   */
  rawImplsMap = {} as Record<
    string,
    { impl: NodeImplConstructor<SerializableNode>; init: SerializableNode }
  >;

  /**
   * A set containing all distinct node types that have been registered. This set is used to quickly check
   * the existence of a node type within the system without needing to search through the complete registry.
   * It supports operations where type checking is necessary, providing a fast access point to determine
   * what types of nodes are available.
   */
  readonly nodeTypes = new Set<NodeTypes>();

  /**
   * A dictionary where each key is a node type and the value is an array of subtypes associated with that type.
   * This structure is crucial for systems that require knowledge of the hierarchical organization of node types,
   * allowing them to access all subtypes related to a specific node type efficiently. It supports operations
   * such as dynamic node creation and validation by providing a clear map of how types are structured and related.
   */
  readonly nodeTypePairs = {} as { [P in NodeTypes]: NodeSubTypes[] };

  /**
   * Retrieves a list of all node implementation constructors registered in the system.
   * This list allows for dynamic instantiation of nodes where only the type is known at runtime.
   * @returns An array of node implementation constructors, enabling the creation of node instances dynamically.
   */
  get nodeConstructors(): NodeImplConstructor<SerializableNode>[] {
    return Object.values(this.implsMap).map((nodeImpl) => nodeImpl.impl);
  }

  /**
   * Provides access to a map of dynamically registered node implementations based on a composite key of `type-subType`.
   * This map is essential for retrieving specific node implementations dynamically,
   * facilitating operations such as node instantiation and processing based on type and subtype.
   * @returns A map where each key is a string in the format `type-subType` and each value is an object containing the node implementation constructor and the initial node instance template.
   */
  get dynamicImplsMap(): Record<
    string,
    { impl: NodeImplConstructor<SerializableNode>; init: SerializableNode }
  > {
    return this.implsMap;
  }

  /**
   * Provides access to a map of dynamically registered node implementations based on the serializable ID of the node.
   * This map is particularly useful for instantiating nodes from serialized data where
   * the node type might not be directly known but can be inferred from the ID.
   * @returns A map where each key is a serializable ID and each value is an object containing the node implementation constructor and the initial node instance template, facilitating node creation from serialized forms.
   */
  get dynamicRawImplsMap(): Record<
    string,
    { impl: NodeImplConstructor<SerializableNode>; init: SerializableNode }
  > {
    return this.rawImplsMap;
  }

  /**
   * Registers a new node implementation into the system, associating it with its type and subtype.
   * It updates the internal mappings to facilitate node creation and identification based on type and subtype.
   * @param impl The node implementation constructor to register.
   * @param registerArgs Optional arguments for initializing the node.
   * @returns An updated instance of `NodeRegistration` reflecting the new node registration.
   */
  register<T extends SerializableNode>(
    impl: NodeImplConstructor<T>,
    registerArgs?: Record<string, unknown>
  ): NodeRegistration<
    NodeTypes | T['type'],
    NodeSubTypes | T['subType'],
    Nodes | T
  > {
    const newRegistration = this as NodeRegistration<
      NodeTypes | T['type'],
      NodeSubTypes | T['subType'],
      Nodes | T
    >;

    const node = impl.create(registerArgs);
    const type = node.type as T['type'];
    const subType = node.subType as T['subType'];

    const key = [type, subType].join('-');

    if (newRegistration.info[key]) {
      throw new Error(`Duplicate node type pair: ${key}`);
    }

    newRegistration.info[key] = {
      impl: impl as any,
    };

    newRegistration.implsMap[key] = {
      impl: impl as any,
      init: node,
    };

    const serializableId = node.data._id.join('-');
    if (!newRegistration.rawImplsMap[serializableId]) {
      newRegistration.rawImplsMap[serializableId] = {
        impl: impl as any,
        init: node,
      };
    }

    newRegistration.nodeTypes.add(type);

    if (newRegistration.nodeTypePairs[type]) {
      newRegistration.nodeTypePairs[type].push(subType);
    } else {
      newRegistration.nodeTypePairs[type] = [subType];
    }

    return newRegistration;
  }

  /**
   * Creates an instance of a node based on the specified type and subtype,
   * using optional arguments for initialization.
   *
   * @param type The primary type of the node.
   * @param subType The subtype of the node.
   * @param registerArgs Optional arguments for node initialization.
   * @returns A new node instance of the specified type and subtype.
   * @throws an error if the type-subtype pair is not registered.
   */
  create<T extends NodeTypes, U extends NodeSubTypes>(
    type: T,
    subType: U,
    registerArgs?: Record<string, unknown>
  ): Extract<Nodes, { type: T; subType: U }> {
    const key = [type, subType].join('-');

    const nodeImpl = this.info[key] as {
      impl: NodeImplConstructor<Extract<Nodes, { type: T; subType: U }>>;
    };

    if (!nodeImpl) {
      throw new Error(`Unknown node type pair: ${type}-${subType}`);
    }

    return nodeImpl.impl.create(registerArgs);
  }

  /**
   * Instantiates a node implementation for a given node object.
   *
   * @param node The node instance to create an implementation for.
   * @returns A new node implementation instance.
   * @throws an error if the node's type-subtype pair is unknown.
   */
  createImpl<T extends Nodes>(node: T): NodeImpl<T> {
    const key = [node.type, node.subType].join('-');

    const nodeImpl = this.info[key] as {
      impl: NodeImplConstructor<
        Extract<Nodes, { type: T['type']; subType: T['subType'] }>
      >;
    };

    if (!nodeImpl) {
      throw new Error(`Unknown node type pair: ${key}`);
    }

    const { impl: Impl } = nodeImpl;

    const newNodeImpl = new Impl(node as any) as unknown as NodeImpl<T>;

    if (!newNodeImpl) {
      throw new Error(`Unknown node type pair: ${key}`);
    }

    return newNodeImpl;
  }

  /**
   * Instantiates a node implementation for a given node object. The serializable data is re-loaded everytime.
   *
   * @param node The node instance to create an implementation for.
   * @returns A new node implementation instance.
   * @throws an error if the node's type-subtype pair is unknown.
   */
  async createLoadImpl<T extends Nodes>(node: T): Promise<NodeImpl<T>> {
    const key = [node.type, node.subType].join('-');

    const nodeImpl = this.info[key] as {
      impl: NodeImplConstructor<
        Extract<Nodes, { type: T['type']; subType: T['subType'] }>
      >;
    };

    if (!nodeImpl) {
      throw new Error(`Unknown node type pair: ${key}`);
    }

    const { impl: Impl } = nodeImpl;

    const data = await load<Serializable>(JSON.stringify(node.data), globalSecretMap, globalImportMap);
    node.data = data;

    const newNodeImpl = new Impl(node as any) as unknown as NodeImpl<T>;

    if (!newNodeImpl) {
      throw new Error(`Unknown node type pair: ${key}`);
    }

    return newNodeImpl;
  }

  /**
   * Dynamically creates an instance of a node based on the specified type and subtype.
   * This method allows for node creation using dynamic type information, which is useful in scenarios where node types and subtypes are determined at runtime.
   * @param type The type of the node to create.
   * @param subType The subtype of the node to create.
   * @param registerArgs Optional arguments for initializing the node.
   * @throws An error if no implementation is registered for the specified type and subtype.
   * @returns A new instance of the specified node type and subtype.
   */
  createDynamic(
    type: string,
    subType: string,
    registerArgs?: Record<string, unknown>
  ): SerializableNode {
    const implClass = this.dynamicImplsMap[`${type}-${subType}`];
    if (!implClass) {
      throw new Error(
        `createDynamic: Unknown node - type: ${type}, subType: ${subType}`
      );
    }

    return implClass.impl.create(registerArgs);
  }

  /**
   * Dynamically creates a node instance from a raw serializable object.
   * This method is particularly useful when nodes need to be instantiated from serialized data or when node types are not known until runtime.
   * @param serializable A raw serializable object that represents the node's data.
   * @param registerArgs Optional arguments for initializing the node.
   * @throws An error if no implementation is found for the node's serializable ID.
   * @returns A new node instance initialized from the provided serializable data.
   */
  createDynamicRaw(
    serializable: Serializable,
    registerArgs?: Record<string, unknown>
  ): SerializableNode {
    const implClass = this.dynamicRawImplsMap[serializable._id.join('-')];
    if (!implClass) {
      throw new Error(
        `createDynamicRaw: Unknown node - name: ${
          serializable._id[serializable._id.length - 1]
        }`
      );
    }

    return implClass.impl.nodeFrom(serializable, registerArgs);
  }

  /**
   * Creates a node implementation dynamically based on the type and subtype of the node.
   * This method is used to instantiate node implementations that are needed for node processing but are determined dynamically based on node data.
   * @param node The node for which the implementation needs to be created.
   * @throws An error if no implementation is registered for the node's type and subtype.
   * @returns An instance of `NodeImpl` for the given node, enabling further interaction and processing.
   */
  createDynamicImpl(node: SerializableNode): NodeImpl<SerializableNode> {
    const { type, subType } = node;
    const implClass = this.dynamicImplsMap[`${type}-${subType}`];
    if (!implClass) {
      throw new Error(
        `createDynamicImpl: Unknown node - type: ${type}, subType: ${subType}`
      );
    }

    const newNodeImpl = new implClass.impl(node);
    if (!newNodeImpl) {
      throw new Error(
        `createDynamicImpl: Unknown node - type: ${type}, subType: ${subType}`
      );
    }

    return newNodeImpl;
  }

  /**
   * Creates a node implementation dynamically based on the type and subtype of the node. The serializable data is re-loaded everytime.
   * This method is used to instantiate node implementations that are needed for node processing but are determined dynamically based on node data.
   * @param node The node for which the implementation needs to be created.
   * @throws An error if no implementation is registered for the node's type and subtype.
   * @returns An instance of `NodeImpl` for the given node, enabling further interaction and processing.
   */
  async createDynamicLoadImpl(node: SerializableNode): Promise<NodeImpl<SerializableNode>> {
    const { type, subType } = node;
    const implClass = this.dynamicImplsMap[`${type}-${subType}`];
    if (!implClass) {
      throw new Error(
        `createDynamicImpl: Unknown node - type: ${type}, subType: ${subType}`
      );
    }

    const data = await load<Serializable>(JSON.stringify(node.data), globalSecretMap, globalImportMap);
    node.data = data;

    const newNodeImpl = new implClass.impl(node);
    if (!newNodeImpl) {
      throw new Error(
        `createDynamicImpl: Unknown node - type: ${type}, subType: ${subType}`
      );
    }

    return newNodeImpl;
  }

  /**
   * Determines if a type-subtype pair is registered in the system.
   * @param type The node type to check.
   * @param subType The node subtype to check.
   * @returns `true` if the type-subtype pair is registered, otherwise `false`.
   */
  isRegistered(type: string, subType: string): boolean {
    const key = [type, subType].join('-');

    return this.info[key] !== undefined;
  }
}

/**
 * Registers all built-in nodes with a newly instantiated `NodeRegistration`.
 * This function centralizes the registration of all predefined nodes.
 * @param registry The `NodeRegistration` instance to register nodes with.
 * @returns The `NodeRegistration` instance after all built-in nodes have been registered.
 */
export function registerBuiltInNodes(registry: NodeRegistration) {
  return registry
    .register(PDFLoaderNodeImpl)
    .register(ChatMessageNodeImpl)
    .register(HumanMessageNodeImpl)
    .register(BotMessageNodeImpl)
    .register(SystemMessageNodeImpl)
    .register(FunctionMessageNodeImpl)
    .register(StringPromptNodeImpl)
    .register(ChatPromptNodeImpl)
    .register(TextSplitterNodeImpl)
    .register(RecursiveTextSplitterNodeImpl)
    .register(LanguageTextSplitterNodeImpl, { language: 'cpp' })
    .register(LanguageTextSplitterNodeImpl, { language: 'go' })
    .register(LanguageTextSplitterNodeImpl, { language: 'java' })
    .register(LanguageTextSplitterNodeImpl, { language: 'js' })
    .register(LanguageTextSplitterNodeImpl, { language: 'php' })
    .register(LanguageTextSplitterNodeImpl, { language: 'proto' })
    .register(LanguageTextSplitterNodeImpl, { language: 'python' })
    .register(LanguageTextSplitterNodeImpl, { language: 'rst' })
    .register(LanguageTextSplitterNodeImpl, { language: 'ruby' })
    .register(LanguageTextSplitterNodeImpl, { language: 'rust' })
    .register(LanguageTextSplitterNodeImpl, { language: 'scala' })
    .register(LanguageTextSplitterNodeImpl, { language: 'markdown' })
    .register(LanguageTextSplitterNodeImpl, { language: 'latex' })
    .register(LanguageTextSplitterNodeImpl, { language: 'html' })
    .register(LanguageTextSplitterNodeImpl, { language: 'sol' })
    .register(TokenTextSplitterNodeImpl)
    .register(OpenAINodeImpl)
    .register(GeminiNodeImpl)
    .register(OpenAIChatNodeImpl)
    .register(GeminiChatNodeImpl)
    .register(VariableValidatorNodeImpl)
    .register(InputNodeImpl)
    .register(IfConditionNodeImpl);
}

/**
 * The `globalNodeRegistry` is a global singleton instance that holds registrations
 * for all the built-in node types within the system. It facilitates the dynamic creation,
 * configuration, and management of nodes based on predefined implementations. This registry
 * is utilized to instantiate nodes directly from serialized data or via type and subtype specifications,
 * ensuring that node creation is consistent and follows the defined specifications.
 *
 * @example
 * #### Creating a Node Directly
 * This example shows how to create a node directly using the registry:
 * ```typescript
 * const pdfLoader = new PDFLoader();
 * const chatNode = globalNodeRegistry.createDynamicRaw(pdfLoader);
 * ```
 *
 * #### Using Node in a Type-Safe Manner
 * For type-safe node creation, particularly when the node type and subtype are known:
 * ```typescript
 * const pdfNode = globalNodeRegistry.createDynamic('pdf', 'loader');
 * ```
 *
 * #### Resetting the Registry
 * In scenarios such as testing or when starting a new session where a clean registry state is needed:
 * ```typescript
 * await resetNodeRegistry();
 * ```
 */
let globalNodeRegistry = registerBuiltInNodes(new NodeRegistration());

export { globalNodeRegistry };

/**
 * Represents the type for all nodes that are registered by default within the system. This type is derived
 * from the nodes registered in the global node registry. It encompasses a variety of node types, each tailored
 * to specific functionalities within the system, ensuring that any operation or transformation that nodes
 * might undergo can be strongly typed.
 */
export type BuiltInNodes = typeof globalNodeRegistry.registeredNodes;

/**
 * Represents the distinct node types that are available within the built-in registry. This type is useful for
 * operations that require a reference to the types of nodes without needing to know the specifics of their
 * implementation or internal structure. It enhances type safety by ensuring that only valid node types are
 * referenced within typed operations.
 */
export type BuiltInNodeTypes = typeof globalNodeRegistry.registeredNodeTypes;

/**
 * Represents the mapping of each built-in node type to its corresponding subtypes. This type is crucial for
 * understanding and accessing the full structure of node types within the system. It is used in contexts where
 * operations depend on the subtype of a node, such as during node creation or when performing type-specific
 * logic.
 */
export type BuiltInNodeTypePairs =
  typeof globalNodeRegistry.registeredNodeTypePairs;

/**
 * A utility type that extracts nodes of a specific type from the built-in nodes based on the provided node type.
 * This type is particularly useful when you need to obtain a subset of nodes that match a given type, allowing
 * for operations that are specific to that type while maintaining type safety.
 *
 * @template T - The type of the node to extract from the built-in nodes.
 */
export type NodeOf<T extends BuiltInNodeTypes> = Extract<
  BuiltInNodes,
  { type: T }
>;

/**
 * Resets the global node registry by re-initializing and re-registering all built-in nodes.
 * Useful for resetting the state in environments where node configurations may change dynamically.
 */
export function resetNodeRegistry() {
  globalNodeRegistry = registerBuiltInNodes(new NodeRegistration());
}
