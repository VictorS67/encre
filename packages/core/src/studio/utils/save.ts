import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'yaml';

import { BaseGraph } from '../graph.js';

/**
 * Serializes a BaseGraph object and saves it as a YAML file. This function
 * converts the graph object into a JSON string, then translates this JSON
 * into YAML format for better readability and stores it in a file.
 *
 * @param graph The BaseGraph instance to be saved.
 * @param cwd The directory where the graph file will be saved. Defaults to the current directory.
 *
 * ### Usage Example:
 * ```typescript
 * const myGraph = new SubGraoh(...);
 *
 * // Save the graph to the default directory
 * await saveGraph(myGraph);
 *
 * // Save the graph to a specific directory
 * await saveGraph(myGraph, '/path/to/save');
 * ```
 *
 * ### Notes:
 * - The file name is derived from the graph's title or defaults to 'graph' if no title is provided.
 * - The file extension used is `.encre` to denote an encoded or serialized file format, potentially specific to the application.
 * - If an error occurs during file writing, it will be logged to the console.
 */
export const saveGraph = async (graph: BaseGraph, cwd: string = './') => {
  try {
    const filePath: string = path.join(
      cwd,
      `${graph.title ?? 'graph'}.encre`
    );

    const serializedStr: string = JSON.stringify(
      await graph.serialize(),
      null,
      2
    );
    const data: string = yaml.stringify(JSON.parse(serializedStr));

    fs.writeFileSync(filePath, data);
  } catch (e) {
    console.error(e);
  }
};
