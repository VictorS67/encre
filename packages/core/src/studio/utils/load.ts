import { type FileLoader, loadFromFile } from '../../utils/load.js';
import { parseFile } from '../../utils/parse.js';
import { BaseGraph } from '../graph.js';
import { type GuardrailRegistration } from '../registration/guardrails.js';
import { type NodeRegistration } from '../registration/nodes.js';

/**
 * Asynchronously loads a graph from a specified file. This function is designed to read
 * a graph's serialized data from a file, deserialize it, and instantiate a `BaseGraph` object.
 * 
 * @param file A string that represents the name or identifier of the file containing the graph data.
 * @param path The path where the file is located.
 * @param values Optional. An object containing key-value pairs that may be required for initializing
 *               nodes within the graph based on the serialized data.
 * @param registry Optional. An object containing `NodeRegistration` for node types and
 *                 `GuardrailRegistration` for guardrails that may be needed during deserialization.
 * @returns A promise that resolves to an instance of `BaseGraph`.
 * 
 * @example
 * ```typescript
 * const graph = await loadGraphFromFile('myApp.encre', '/graphs');
 * ```
 */
export const loadGraphFromFile: FileLoader<BaseGraph> = async (
  file: string,
  path: string,
  values: Record<string, unknown> = {},
  registry?: {
    nodes?: NodeRegistration;
    guardrails?: GuardrailRegistration;
  }
): Promise<BaseGraph> => {
  const serialized = parseFile(file, path);
  return BaseGraph.deserialize(serialized, values, registry);
};

/**
 * Asynchronously loads a graph from a URI, typically a path to a file. This function wraps around
 * `loadGraphFromFile`, allowing for indirect graph loading through a URI, facilitating the separation
 * of graph loading logic from file handling.
 * 
 * @param uri A string URI pointing to the location of the graph data. This could be a file path or URL.
 * @param values Optional. An object containing key-value pairs that may be required for initializing
 *               nodes within the graph based on the serialized data.
 * @param registry Optional. An object containing `NodeRegistration` for node types and
 *                 `GuardrailRegistration` for guardrails that may be needed during deserialization.
 * @returns A promise that resolves to an instance of `BaseGraph`.
 * 
 * @example
 * ```typescript
 * const graph = await loadGraph('/api/graphs/myGraph');
 * ```
 */
export const loadGraph = async (
  uri: string,
  values: Record<string, unknown> = {},
  registry?: {
    nodes?: NodeRegistration;
    guardrails?: GuardrailRegistration;
  }
): Promise<BaseGraph> => {
  return loadFromFile(uri, loadGraphFromFile, values, registry);
};
