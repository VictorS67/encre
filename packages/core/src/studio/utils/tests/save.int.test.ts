import path from 'path';
import url from 'url';
import { describe, expect, test } from '@jest/globals';

import { OpenAIChat } from '../../../events/inference/chat/llms/openai/chat.js';
import { StringPrompt } from '../../../events/input/load/prompts/text.js';
import { SubGraph } from '../../graph.js';
import { globalNodeRegistry } from '../../registration/nodes.js';
import { loadGraph } from '../load.js';
import { saveGraph } from '../save.js';

describe('test saving to local', () => {
  test('save to path', async () => {
    const stringPromptNode = globalNodeRegistry.createDynamicRaw(
      new StringPrompt('Who are you?')
    );

    const openAINode = globalNodeRegistry.createDynamicRaw(
      new OpenAIChat({ modelName: 'gpt-3.5-turbo' })
    );

    const graph = new SubGraph({
      nodes: [stringPromptNode, openAINode],
      connections: [
        {
          fromNodeId: stringPromptNode.id,
          fromPortName: 'prompt',
          toNodeId: openAINode.id,
          toPortName: 'prompt',
        },
      ],
    });

    const filePath: string = path.resolve(
      path.dirname(url.fileURLToPath(import.meta.url)),
      './examples/'
    );

    await saveGraph(graph, filePath);

    expect(true).toBeTruthy();
  });

  test('load graph', async () => {
    const filePath: string = path.resolve(
      path.dirname(url.fileURLToPath(import.meta.url)),
      './examples/Untitled Graph.encre'
    );

    const graph = await loadGraph(filePath);

    expect(graph).toMatchSnapshot();
  });
});
