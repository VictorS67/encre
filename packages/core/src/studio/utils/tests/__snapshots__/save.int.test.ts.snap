// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test saving to local load graph 1`] = `
{
  "_grp": 1,
  "_id": [
    "studio",
    "graph",
    "SubGraph",
  ],
  "_kwargs": {
    "connections": [
      {
        "fromNodeId": "@vAJbPalSsu5XkepR",
        "fromPortName": "prompt",
        "toNodeId": "D2CixRBhh4Iu_J4Xz",
        "toPortName": "prompt",
      },
    ],
    "description": "",
    "id": "6ot@YhxtGapyiMhkQ",
    "nodes": [
      {
        "data": {
          "_grp": 1,
          "_id": [
            "events",
            "input",
            "load",
            "prompts",
            "text",
            "StringPrompt",
          ],
          "_kwargs": {
            "value": {
              "value": "Who are you?",
            },
          },
          "_type": "constructor",
        },
        "id": "@vAJbPalSsu5XkepR",
        "inputs": {},
        "memory": 0,
        "outputSizes": {
          "prompt": 0,
        },
        "outputs": {
          "prompt": [
            "string",
            "chat-message[]",
          ],
        },
        "registerArgs": undefined,
        "runtime": 0,
        "subType": "string",
        "type": "prompt",
        "visualInfo": {
          "position": {
            "x": 0,
            "y": 0,
          },
          "size": {
            "height": 500,
            "width": 300,
          },
        },
      },
      {
        "data": {
          "_grp": 1,
          "_id": [
            "events",
            "inference",
            "chat",
            "chatlms",
            "openai",
            "OpenAIChat",
          ],
          "_kwargs": {
            "additional_kwargs": {},
            "callbacks": undefined,
            "frequency_penalty": 0,
            "logit_bias": undefined,
            "max_tokens": 2048,
            "model": "gpt-3.5-turbo",
            "openai_api_key": {
              "_grp": 1,
              "_id": [
                "OPENAI_API_KEY",
              ],
              "_type": "secret",
            },
            "presence_penalty": 0,
            "response_format_type": undefined,
            "seed": undefined,
            "stop_words": undefined,
            "stream": false,
            "temperature": 1,
            "timeout": undefined,
            "top_p": 1,
            "user": undefined,
            "verbose": undefined,
          },
          "_type": "constructor",
        },
        "id": "D2CixRBhh4Iu_J4Xz",
        "inputs": {
          "prompt": [
            "string",
            "chat-message[]",
          ],
        },
        "memory": 0,
        "outputSizes": {
          "info": 0,
          "message": 0,
          "output": 0,
          "tokenUsage": 0,
        },
        "outputs": {
          "info": [
            "object",
            "unknown",
          ],
          "message": "chat-message",
          "output": "string",
          "tokenUsage": [
            "object",
            "unknown",
          ],
        },
        "registerArgs": undefined,
        "runtime": 0,
        "subType": "openai",
        "type": "chatlm",
        "visualInfo": {
          "position": {
            "x": 0,
            "y": 0,
          },
          "size": {
            "height": 500,
            "width": 300,
          },
        },
      },
    ],
    "registry": undefined,
    "title": "Untitled Graph",
  },
  "_type": "constructor",
}
`;
